File: 01.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script

{-# OPTIONS_GHC -Wall #-}

import Data.Maybe (fromJust)
import Text.ParserCombinators.ReadP
import Utils

main :: IO ()
main = do
    input <- readFile "01.input"
    let shifts = parseMany (int <* eof) $ lines input
    -- part 1
    print $ sum shifts
    -- part 2
    print $ (fromJust . firstDup . scanl (+) 0 . cycle) shifts

File: 01.py
#!/usr/bin/env python3

from itertools import cycle

from utils import first_dup


def shifted(shifts, start=0):
    yield start
    for s in shifts:
        start += s
        yield start


if __name__ == '__main__':
    with open('01.input') as f:
        shifts = [int(line) for line in f]

    # part 1
    print(sum(shifts))

    # part 2
    print(first_dup(shifted(cycle(shifts))))
